Sources/SwiftFormatWhitespaceLinter/WhitespaceLinter.swift

```diff
@@ -0,0 +1,332 @@
+    )
+  }
+
+  /// Find the next non-whitespace character in a given string, and any leading whitespce before the
+  /// character.
+  ///
+  /// If the character at `offset` is whitespace, we scan forward until we find a non-whitespace
+  /// character. We then return the new offset, the character we landed on, and a string containing
+  /// the character's leading whitespace.
+  ///
+  /// - Parameters:
+  ///   - offset: The printable character offset within the string.
+  ///   - data: The input string.
+  /// - Returns a tuple of the new offset, the non-whitespace character we landed on, and a string
+  ///   containing the leading whitespace.
+  func nextCharacter(offset: Int, data: String)
+  -> (offset: Int, char: Character?, whitespace: String) {
+    var whitespaceBuffer = ""
+
+    for i in offset..<data.count {
+      let index = data.index(data.startIndex, offsetBy: i)
+      let char = data[index]
+
+      if [" ", "\n"].contains(char) {
+        whitespaceBuffer += String(char)
+      }
+      else {
+        return (offset: i, char: char, whitespace: whitespaceBuffer)
+      }
+    }
+    return (offset: data.count - 1, char: nil, whitespace: whitespaceBuffer)
+  }
+
+  /// Given a string and a printable charater offset, calculate the line and column number.
+  ///
+  /// - Parameters:
+  ///   - offset: The printable character offset.
+  ///   - data: The input string for which we want the line and column numbers.
+  /// - Returns a tuple with the line and column numbers within `data`.
+  func calculatePosition(offset: Int, data: String) -> (line: Int, column: Int) {
+    var line = 1
+    var column = 0
+
+    for (index, char) in data.enumerated() {
+      if char == "\n" {
+        line += 1
+        column = 0
+      }
+      else {
+        column += 1
+      }
+      if index == offset { break }
+    }
+    return (line: line, column: column)
+  }
+
+  /// Emits the provided diagnostic message to the DiagnosticEngine. The message will correspond to
+  /// a specific location (line and column number) in the input Swift source file (`userText`).
+  ///
+  /// - Parameters:
+  ///   - message: The Diagnostic.Message object we wish to emit.
+  ///   - line: The line number location of the message
+  ///   - column: The column number location of the message
+  ///   - utf8Offset: The utf8 offset location of the message
+  ///   - actions: Used for attaching notes, highlights, etc.
+  func diagnose(
+    _ message: Diagnostic.Message,
+    line: Int,
+    column: Int,
+    utf8Offset: Int,
+    actions: ((inout Diagnostic.Builder) -> Void)? = nil
+  ) {
+    let loc = SourceLocation(
+      file: context.fileURL.path,
+      position: AbsolutePosition(line: line, column: column, utf8Offset: utf8Offset)
+    )
+    context.diagnosticEngine?.diagnose(
+      message,
+      location: loc,
+      actions: actions
+    )
+  }
+}
+
+extension Diagnostic.Message {
+  static func trailingWhitespaceError() -> Diagnostic.Message {
```

Comment:
For this one and `removeLineError`, you can make them `static let` properties instead of functions with no arguments.

Reply:
Done.

Reply:
It looks like this one didn't get changed?

Reply:
Oops. I got `removeLineError`, but missed this one. Should be fixed now.